File: src-py/nlp/knn.py
"""
Implementation of k-nearest neighbours classifier
"""

import numpy as np
from scipy import stats
import utils

class KNN:

    def __init__(self, k):
        self.k = k

    def fit(self, X, y):
        self.X = X # just memorize the trianing data
        self.y = y

    def predict(self, Xtest):
        X = self.X
        y = self.y
        n = X.shape[0]
        t = Xtest.shape[0]
        k = min(self.k, n)

        # Compute cosine_distance distances between X and Xtest
        dist2 = cosine_distance(X, Xtest)

        # yhat is a vector of size t with integer elements
        yhat = np.ones(t, dtype=np.uint8)
        for i in range(t):
            # sort the distances to other points
            inds = np.argsort(dist2[:,i])

            # compute mode of k closest training pts
            yhat[i] = stats.mode(y[inds[:k]])[0][0]

        return yhat

     def cosine_distance(X1,X2):
         pass
         "Insert your code here"

File: src-cs/Classifier/KNN.cs
using System;
using Main;

namespace Main.Classifier
{
    class KNN
    {
        private int k;
        private double[,] x;
        private int[] y;
        public KNN(int k)
        {
            this.k = k;
        }

        public void fit(double[,] x, int[] y)
        {
            this.x = x;
            this.y = y;
        }

        public int[] Predict(double[,] x)
        {
            int t = this.x.GetLength(0);
            int n = x.GetLength(0);
            int d = x.GetLength(1);
            int[] retval = new int[n];
            // for each data example
            for (int i = 0; i < n; i++)
            {
                // for each point in the model
                double distances = new double[t];
                for (int c = 0; c < t; c++)
                {
                    distances[c] = Util.Distance(x[i], this.x[c]);
                }
                int[] nearest = Util.MinArg(distances, k);
                retval[i] = Util.Mode(nearest);
            }
            return retval;
        }
    }
}
